lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -4245,6 +4245,96 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
     return BM->addInstTemplate(internal::OpMaskedScatterINTEL, Ops, BB,
                                nullptr);
   }
+  case Intrinsic::is_fpclass: {
+    // There is no direct counterpart for the intrinsic in SPIR-V, hence
+    // we need to emulate its work by sequence of other instructions
+    SPIRVType *Ty = transType(II->getType());
+    auto *InputFloat = transValue(II->getArgOperand(0), BB);
+    uint64_t FPClass =
+        cast<ConstantInt>(II->getArgOperand(1))->getZExtValue();
+    std::vector<SPIRVValue *> ResultVec;
+    std::vector<SPIRVId> Ops = {InputFloat->getId()};
+    // Adds test for Negative/Positive values
+    auto GetNegPosInstTest = [&](SPIRVValue *TestInst, bool IsNegative) {
+      auto *SignBitTest =
+          BM->addInstTemplate(OpSignBitSet, {TestInst->getId()}, BB, Ty);
+      if (IsNegative)
+        return SignBitTest;
+      return BM->addInstTemplate(OpLogicalNot, {SignBitTest->getId()}, BB, Ty);
+    };
+    // Integer parameter of the intrinsic is combined from several bit masks
+    // referenced in FPClassTest enum from FloatingPointMode.h in LLVM.
+    // Since a single intrinsic can provide multiple tests - here we might end
+    // up adding several sequences of SPIR-V instructions
+    if (FPClass & fcSNan || FPClass & fcQNan) {
+      // There are no signaling and quite NaN instructions in SPIR-V, so map
+      // both of the tests on OpIsNan
+      ResultVec.emplace_back(BM->addInstTemplate(OpIsNan, Ops, BB, Ty));
```

Comment:
It's not correct to consider the tests for sNaN or qNaN as equivalent, because the definition is based on the bit pattern (and InstCombine may canonicalize such a test to an fpclass check!). Distinguishing between qNaN and sNaN requires checking if the quiet bit (MSB of the mantissa) is set or not.